### PR TITLE
[apps] improved detection app

### PR DESCRIPTION
**Pull request by simogasp**
Just small improvements to the detection app to be able to visualize the detected markers using opencv and optionally to save the result with the graphical overlay.

* it introduces the option `--save-detected-image` to save the images with the detected markers, displaying the ID and the outer ellipse.

* the option `--show-unreliable` is more for debugging purposes and allows to display the false detection as well

The graphical overlay looks like this (using images from ./sample):

01_detected

02_detected

### CODE REVIEW

File: src/applications/detection/main.cpp [hunk 1]
```diff
@@ -77,16 +77,19 @@ void drawMarkers(const boost::ptr_list<CCTag> &markers, cv::Mat &image)
     if(marker.getStatus() == status::id_reliable)
     {
       const cv::Scalar color = cv::Scalar(0, 255, 0 , 255);
+        const auto rescaledOuterEllipse = marker.rescaledOuterEllipse();
+        std::cout << "a " << rescaledOuterEllipse.a() << " b " << rescaledOuterEllipse.b() << " angle " << rescaledOuterEllipse.angle() << std::endl;
```
Comment (griwodz) [inline src/applications/detection/main.cpp:81]: Do you want to keep the cout line for information about the ellipse?

File: src/applications/detection/main.cpp [hunk 2]
```diff
@@ -481,10 +518,26 @@ int main(int argc, char** argv)
           int pipeId = (fileInFolder.first & 1);
           boost::ptr_list<CCTag> markers;
 #ifdef PRINT_TO_CERR
-          detection(fileInFolder.first, pipeId, imgGray, params, bank, markers, std::cerr, fileInFolder.second.stem().string());
+          detection(fileInFolder.first, pipeId, imgGray, params, bank, markers, std::cerr, inputImageFile.stem().string());
 #else
           detection(fileInFolder.first, pipeId, imgGray, params, bank, markers, outputFile, fileInFolder.second.stem().string());
```
Comment (griwodz) [inline src/applications/detection/main.cpp:523]: Use inputImageFile in the else case as well?

**Review by griwodz — CHANGES_REQUESTED** [review]
This looks good to me, I have only added two tiny change requests.

File: src/applications/detection/main.cpp (hunk 1 above)
Comment (simogasp) [reply src/applications/detection/main.cpp:81]: nope, it was just debugging stuff, i'll remove it.

File: src/applications/detection/main.cpp (hunk 2 above)
Comment (simogasp) [reply src/applications/detection/main.cpp:523]: yes, bad refactoring...

File: src/applications/detection/main.cpp (hunk 1 above)
Comment (simogasp) [reply src/applications/detection/main.cpp]: d31f2d0

File: src/applications/detection/main.cpp (hunk 2 above)
Comment (simogasp) [reply src/applications/detection/main.cpp]: 9c817cd

**Comment by simogasp** [comment]
before merging it i\d like to do a code format of the main.cpp as there are different styles, a mix of tabs and space etc.
I didn't do it otherwise it would have been a mess for the review, but once it is good i can do a pass before merging

**Comment by griwodz** [comment]
Build errors. Is the Trusty environment too old?
Can the ::gets() problem for clang be fixed by including <stdio.h> before the opencv headers?

**Comment by griwodz** [comment]
We are using Ubuntu Xenial in Popsift, perhaps its OK to require that for CCTag?

**Comment by simogasp** [comment]
we cannot use trusty anymore because of c++14, switching to xenial requires rebuilding all the dependencies (opencv etc) from our repo AlicevisionDependencies.  Popsift still works because it only depends (optionally) on boost and we used the system one on travis.
I think github action is more maintainable because of the docker images, I guess we can also use the docker images on travis but I don't know it is worth it to maintain two different things. We can for sure add other tests on github actions using different images with different versions of cuda

**Comment by simogasp** [comment]
yesterday I actually tried to fix it (https://github.com/alicevision/CCTag/commit/27905876329319556b4257facf58401b07bdc146) but I've forgotten about the way we use the pre-built dependencies in AlicevisionDependencies.

**Pull request merged**